### ISSUE TITLE
v1.47.1 much slower

### ISSUE THREAD

**Issue by fredbi** [golangci/golangci-lint#2997]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).
- [X] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

### Description of the problem

Today, our CI pipelines moved to v1.47.1 and we immediately felt the pain:
linting was suddenly taking about 3x longer, exhausting the resources of the CI VM (2 vCPU, 4GB RAM), even with a very small amount of changes and the "--new-from-rev master" flag. This behavior was consistent and observed on a dozen PRs today, even those with no change in the go code base.
run is slower when compared to v1.46.2, which was our previous version.

Running v1.47.1 from docker image: golangci/golangci-lint:latest-alpine
Extra installations required for alpine:
apk add git openssh-client bash ncurses gcc musl-dev

CLI for linting:
 golangci-lint run --timeout 10m --new-from-rev master --build-tags integration

### Version of golangci-lint

golangci-lint version running on the CI container.

<details>

```console
$ golangci-lint --version
golangci-lint has version v1.47.1 built from ebd6dcbf on 2022-07-19T11:57:02Z
```

</details>

### Configuration file

<details>

```console
$ cat .golangci.yml
linters-settings:
  govet:
    check-shadowing: true
  gocyclo:
    min-complexity: 25
  maligned:
    suggest-new: true
  dupl:
    threshold: 150
  goconst:
    min-len: 3
    min-occurrences: 2
  forbidigo:
    forbid:
      - ^print.*$
      - 'fmt\.Print.*'
  funlen:
    lines: 100
    statements: 50
  gocognit:
    min-complexity: 80 # This is super high value. We should gradually lower it to 30-40.
run:
  skip-dirs:
    - integrationtest

linters:
  enable:
    - gofmt
    - goimports
    - bodyclose
    - dupl
    - forbidigo
    - gocognit
    - gocritic
    - goimports
    - gosec
    - nakedret
    - nolintlint
    - revive
    - stylecheck
    - unconvert
    - unparam
  disable:
    - maligned
    - lll
    - gochecknoinits
    - gochecknoglobals

```

</details>

### Go environment

go version running on the CI container.

<details>

```console
$ go version && go env
go version go1.18.4 linux/amd64
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/go/pkg/mod"
GONOPROXY="github.com/oneconcern/*"
GONOSUMDB="github.com/oneconcern/*"
GOOS="linux"
GOPATH="/go"
GOPRIVATE="github.com/oneconcern/*"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.18.4"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/root/project/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3438759853=/tmp/go-build -gno-record-gcc-switches"
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
# paste output here
```

</details>

### Code example or link to a public repository

<details>
N/A (private repo)
</details>

**Comment by ldez** [comment]
Hello,

is slower than which version?

**Comment by fredbi** [comment]
Editing description: slower when compared to v1.46.2

**Comment by ChrisSargent** [comment]
Seems to be related to the `revive` linter for use - once we enable that the linter just keeps going and going.....

**Comment by Prajyot-Parab** [comment]
I have noticed the same with (`v1.47.0` & `v1.47.1`), once I disable `revive` linter everything works smooth, with `revive` enabled it seems to take a lot of toll the resources and eventually times out. (checked in `cluster-api-provider-{ibmcloud,azure,gcp}`)

**Comment by Prajyot-Parab** [comment]
I bumped my Golang binary to `v1.18.4` and also the upstream codebase (cluster-api-provider project) Golang to `v1.18.4`, issue got fixed for me, linters working smoothly.

@killianmuldoon ^^^

**Comment by josephlr** [comment]
We also ran into a similar issue when updating to 1.47.1

**Comment by fredbi** [comment]
@Prajyot-Parab 
> I bumped my Golang binary to v1.18.4

My initial report was already running go v1.18.4. This is not likely to be a fix.

**Comment by ldez** [comment]
Can someone give me a public repo?

I tried on several large repos and I don't see any regression (since 1.47.1)

**Comment by killianmuldoon** [comment]
https://github.com/kubernetes-sigs/cluster-api/pull/6943 is the PR on cluster api with the version bump. You can lint locally using `make lint`

**Comment by killianmuldoon** [comment]
This issue is odd - but it's reliably repeatable on the Cluster API repo. As far as I can tell it's a timeout that seems to be caused by revive as used through golangci-lint - we're getting timeouts in all of our CI runs and the same thing is happening locally with golangci-lint 1.47+

Interestingly I don't seem to be able to replicate this when running a debug version of golangci-lint locally, still trying to figure out what the delta is between the `make lint` and debug run configuration.

**Comment by killianmuldoon** [comment]
The issue is occurring in our `make lint` which already uses that script for the install, so I think the root cause is in the binary/config rather than in the action.

**Comment by ldez** [comment]
I can confirm that the issue is with `revive`, I need to investigate more to understand the core of the problem.
Stay tunes.

**Comment by Trim21** [comment]
I try to run it with cpu profiling, And I find most of time is spending on file operator.

first (1487s):
![image](https://user-images.githubusercontent.com/13553903/180019979-47afc59d-9686-44e3-b25a-0f821ce9c702.png)

second (5s):
![image](https://user-images.githubusercontent.com/13553903/180020017-a6791dd3-395f-4182-9b56-310cc4f17f1d.png)

**Comment by ldez** [comment]
In fact, the problem occurs in another process (`exec.Command`), there are a lot of calls to `go list -json -export -- ...`

**Comment by ttys3** [comment]
`the_only_name: 39.849266474s` ?

```
INFO [linters context/goanalysis] analyzers took 48.871097346s with top 10 stages: the_only_name: 39.849266474s, buildir: 1.634334175s, forbidigo: 918.319972ms, gocritic: 443.523491ms, gofumpt: 408.227795ms, goimports: 371.373694ms, dupl: 322.186875ms, directives: 319.432121ms, gofmt: 286.378648ms, gochecknoglobals: 268.964483ms 
```

**Comment by ldez** [comment]
I think it's related to a change inside the package `gcexportdata` from `golang.org/x/tools`:
https://github.com/golang/tools/commit/2a900561e78a69afe5828ff8388aeb0dfc6220dc

`revive` uses `gcexportdata.NewImporter` and the `Importer` uses `gcexportdata.Find`

https://github.com/mgechev/revive/blob/60e1ae4a014ff158513922aff1597ec974945546/lint/package.go#L30

**Comment by ttys3** [comment]
> the_only_name

actually this is not off-topic.

this is the same slow down caused by  revive.  I disabled `revive`, then the speed comese back

why golangci-lint use the name `the_only_name` for revive ?

**Comment by ttys3** [comment]
> `the_only_name: 39.849266474s` ?
> 
> ```
> INFO [linters context/goanalysis] analyzers took 48.871097346s with top 10 stages: the_only_name: 39.849266474s, buildir: 1.634334175s, forbidigo: 918.319972ms, gocritic: 443.523491ms, gofumpt: 408.227795ms, goimports: 371.373694ms, dupl: 322.186875ms, directives: 319.432121ms, gofmt: 286.378648ms, gochecknoglobals: 268.964483ms 
> ```

the_only_name

actually this is not off-topic.

this is the same slow down caused by revive. I disabled revive, then the speed comese back

why golangci-lint use the name the_only_name for revive ?

**Comment by ldez** [comment]
@ttys3 it's offtopic because it's related to this bug.

golangci-lint uses `the_only_name` as a name for the wrapped linter of a linter.

**Comment by rfay** [comment]
Yes, this is very, very painful. Orders of magnitude slower.

**Comment by ldez** [comment]
I opened an issue on `revive` https://github.com/mgechev/revive/issues/713

In summary:
- the problem is related to a change inside `golang.org/x/tools` which impacts performance a lot.
- `revive` use the code impacted by this change.
- inside golangci-lint we use a more recent version of `golang.org/x/tools` than revive.
- we have to use a more recent version of `golang.org/x/tools` to support generics.

**Comment by killianmuldoon** [comment]
Thanks @ldez  - I really appreciate the effort that went into debugging this. I'll take time to assess whether or not to remove revive and re-enable it later. Cluster API is still on 1.44 waiting for a gci fix that came out in this release 😆 Back to drawing board for our upgrade plan!

**Comment by ldez** [comment]
I can create a version without the `revive` rules that have a performance issue.

- `var-declaration` (*default*): https://revive.run/r#var-declaration
- `errorf` (*default*): https://revive.run/r#errorf
- `unexported-return` (*default*): https://revive.run/r#unexported-return
- `time-naming` (*default*): https://revive.run/r#time-naming
- `context-keys-type` (*default*): https://revive.run/r#context-keys-type
- `modifies-value-receiver`: https://revive.run/r#modifies-value-receiver
- `range-val-address`: https://revive.run/r#range-val-address
- `string-of-int`: https://revive.run/r#string-of-int
- `time-equal`: no documentation
- `unhandled-error`: https://revive.run/r#unhandled-error


Question for everybody: is it an acceptable workaround for you?
- yes -> :+1: 
- no -> :-1:

**Comment by killianmuldoon** [comment]
I need to speak with the CAPI maintainers to figure out if this is an acceptable workaround - but I've tried this solution over on the PR to CAPI and it's working 🙂. 

Thanks again for working so quickly on root-causing this issue!

**Comment by ncw** [comment]
I've noticed this problem too in https://github.com/rclone/rclone - linting has gone from a few seconds to 20 minutes!

> I can create a version without the `revive` rules that have a performance issue.

I think that would be a great interim measure. :+1: from me :-)

**Comment by ldez** [comment]
The [proposal](https://github.com/golangci/golangci-lint/issues/2997#issuecomment-1190681977) has been implemented and shipped to a release [v1.47.2](https://github.com/golangci/golangci-lint/releases/tag/v1.47.2)

We will follow and help the revive team to find a fix.

**Comment by silverwind** [comment]
New version seems faster, but it produces log spam, this message 242 times in our case:

```
level=warning msg="[linters context] wastedassign is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649."
```

**Comment by fredbi** [comment]
@ldez thank you so much for the quick workaround. I can confirm that our CI pipelines now resumed to normal behavior. We'll wait quietly until the root cause can be addressed in revive.

**Comment by michaelbeaumont** [comment]
We're still seeing this [issue on `1.48.0`](https://app.circleci.com/pipelines/github/kumahq/kuma/15977/workflows/fc3907ce-12a9-4b59-8709-19ab22e5bdc6/jobs/207051) but [not on `1.47.0` (job failure comes after golangci-lint](https://app.circleci.com/pipelines/github/kumahq/kuma/15978/workflows/9c07b1bd-8b9e-46c9-b590-b4d0fa5ec908/jobs/207057)

**Comment by ldez** [comment]
yes because the fix is not in a release for now.

**Comment by michaelbeaumont** [comment]
I see, I was under the impression the revive check disabling fixed the issue. Got it, thanks!

**Comment by killianmuldoon** [comment]
> yes because the fix is not in a release of golangci-lint for now.
Thanks again for fixing this @ldez. Is there any guideline on how long we might be waiting for a release? We're trying to decide whether to update to 1.48.0 or straight to 1.48.1

**Comment by ldez** [comment]
As we have already merged non-fix PRs, we will not release a v1.48.1.
The next release will be v1.49.0, so you will have to wait a bit.

**Comment by killianmuldoon** [comment]
Thanks! I think we'll update to 1.48.1 for now, and pick up 1.49.0 as soon as it's released.
